transports/onion/src/provider.rs

```diff
@@ -0,0 +1,132 @@
+use std::pin::Pin;
+
+use arti_client::DataStream;
+use futures::{AsyncRead, AsyncWrite};
+
+#[derive(Debug)]
+pub struct OnionStream {
+    inner: DataStream,
```

Comment:
Can we just expose `DataStream` directly? Or does that not work because something further down the stack requires `futures::Async{Read,Write}` even with the `tokio` feature`?

Reply:
I am not happy with these `cfg`s in here. Will need to give this more thought :)

Reply:
> Can we just expose `DataStream` directly? Or does that not work because something further down the stack requires `futures::Async{Read,Write}` even with the `tokio` feature`?

I'm actually not quite sure. Although this part is, along with the cfg's, ported from libp2p-tcp.

Reply:
I had a look at what we currently do in `libp2p-tcp` and I think it is worth doing the same thing here:

1. For `async-io`, we can directly expose `DataStream` because it implements the traits directly, I don't think it is necessary to wrap that again.
2. For `tokio`, we can wrap `DataStream` and use the adaptors to implement `futures::Async{Read,Write}`.

That way, it is clear which types are exposed for which features and enabling all of them is not a problem and does not require any boolean operations on the feature set.

To realize this, we may a `Provider` abstraction, similar to what we do in `libp2p-tcp`. It should be fairly simple though because we only need to adapt what we do with the return value from `TorClient::connect`.